src/cedict-service/indexed-db-store.js

```diff
@@ -0,0 +1,228 @@
+    // DB object is not available when store object is created. It must be added later.
+    this._db = null
+  }
+
+  /**
+   * Checks if the configuration supplied has all the necessary information in it.
+   * If configuration is not valid it will throw an error indicating which check failed.
+   *
+   * @param {object} configuration - A JSON like configuration object.
+   */
+  static checkConfiguration (configuration) {
+    if (!configuration.name) throw new Error('A store name is missing from a configuration')
+    if (!configuration.primaryIndex) throw new Error('A primaryIndex tree is missing from a configuration')
+  }
+
+  /**
+   * Associates a store with an IndexedDB interface of a database where it is located.
+   *
+   * @param {IDBDatabase} db - An interface for connecting to IndexedDB.
+   * @returns {Store} A self-reference for chaining.
+   */
+  associateWith (db) {
+    this._db = db
+    return this
+  }
+
+  /**
+   * Creates a store. Can be run from `onupgradeneeded` callback only.
+   *
+   * @returns {Promise<undefined>|Promise<Error>} A promise that is resolved if store was created successfully
+   *          and is rejected if creation failed.
+   */
+  create () {
+    return new Promise((resolve, reject) => {
+      const options = this._configuration.primaryIndex.keyPath ? { keyPath: this._configuration.primaryIndex.keyPath } : undefined
+      const store = this._db.createObjectStore(this._configuration.name, options)
+      if (this._configuration.indexes) {
+        Object.values(this._configuration.indexes).forEach(idx => {
+          try {
+            store.createIndex(idx.name, idx.keyPath, { unique: idx.unique })
+          } catch (error) {
+            reject(error)
+          }
+        })
+      }
+      resolve()
+    })
+  }
+
+  /**
+   * Deletes all records from the store.
+   *
+   * @returns {Promise<undefined>|Promise<Error>} A promise that is resolved if all records were removed successfully
+   *          and is rejected if operation failed.
+   */
+  clear () {
+    return new Promise((resolve, reject) => {
+      let transaction = this._db.transaction(this._configuration.name, IndexedDbStore.accessModes.READ_WRITE) // eslint-disable-line prefer-const
+      let objectStore = transaction.objectStore(this._configuration.name) // eslint-disable-line prefer-const
+      let clearRequest = objectStore.clear() // eslint-disable-line prefer-const
+      clearRequest.onsuccess = () => resolve()
+      transaction.onerror = (event) => reject(event)
+    })
+  }
+
+  /**
+   * Deletes a store from an IndexedDB database. Can be run from `onupgradeneeded` callback only.
+   *
+   * @returns {Promise<undefined>|Promise<Error>} A promise that is resolved if store was removed successfully
+   *          and is rejected if operation failed.
+   */
+  destroy () {
+    return new Promise((resolve) => {
+      this._db.deleteObjectStore(this._configuration.name)
+      resolve()
```

Comment:
does it need an explicit reject or is that automatic if deleteObjectStore throws an error? (Sorry forever wrapping my head around promises)

Reply:
There is an invisible try/catch block around promise executor functions that converts anything that is thrown into rejections: https://javascript.info/promise-error-handling#implicit-try-catch. I've just learnt of this feature and I think it's very convenient.

Reply:
ok, thanks!